state/machine.go

```diff
@@ -768,16 +779,21 @@ func IsHasAttachmentsError(err error) bool {
 // than the supplied value. If the machine already has that lifecycle
 // value, or a later one, no changes will be made to remote state. If
 // the machine has any responsibilities that preclude a valid change in
-// lifecycle, it will return an error.
-func (original *Machine) advanceLifecycle(life Life, force bool, maxWait time.Duration) (err error) {
-	containers, err := original.Containers()
-	if err != nil {
-		return err
+// lifecycle, it will return an error. dyingAllowContainers indicates
+// whether the machine can have containers when moving to the dying state.
+// Not allowed for moving to dead.
+func (original *Machine) advanceLifecycle(life Life, force, dyingAllowContainers bool, maxWait time.Duration) (err error) {
+	logger.Debugf("%s.advanceLifecycle(%s, %t, %t)", original.Id(), life, force, dyingAllowContainers)
+
+	if life == Dead && dyingAllowContainers {
+		return errors.BadRequestf("life cannot be Dead if dyingAllowContainers true.")
 	}
-	if len(containers) > 0 {
-		return &HasContainersError{
-			MachineId:    original.doc.Id,
-			ContainerIds: containers,
+
+	// A machine can be set to dying with containers, but cannot have any when
+	// advanced to dead.
+	if !dyingAllowContainers && life == Dying {
+		if err := original.evaulateContainersAdvanceLifecycle(); err != nil {
+			return err
```

Comment:
Part of the real change to advanceLifecycle allowed machines with container to be marked dying.